### PR TITLE
Dimension equation theory oracle

### PR DESCRIPTION
**Pull request by favonia**
I need to commit this before switching back to the master branch.

### CODE REVIEW

File: src/basis/DisjointSet.ml [hunk 1]
```diff
@@ -2,41 +2,50 @@ module type S =
...
   type 'a t
 
-  val init : size:int -> 'a t
-  val union : 'a -> 'a -> 'a t -> 'a t
-  val find : 'a -> 'a t -> 'a
-  val find_class : 'a -> 'a t -> 'a list
+  type 'a expr = [ `Const of int | `Var of 'a ]
+
+  val init : size:int -> nconst:int -> 'a t
+  val union : 'a expr -> 'a expr -> 'a t -> [ `Ok of 'a t | `Owned ]
```
Comment (jonsterling) [inline src/basis/DisjointSet.ml]: 😆

**Comment by favonia** [comment]
@jonsterling Do you want to expose the constructors of `I.action`? The oracle needs to know the constructors.

**Comment by jonsterling** [comment]
@favonia Feel free to expose anything you need. We'll be changing those anyway, but for now, just expose it

File: src/core/IDisjointSet.ml [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+  let test x y (h : t) =
+    match query_index x h.index, query_index y h.index with
+    | `Owned, `Owned -> x == y
+    | `Owned, _ -> false
+    | _, `Owned -> false
+    | `Ok x, `Ok y -> find x h == find y h
+
+  let hide (x : Name.t) (h : t) =
+    {h with index = T.remove x h.index}
```
Comment (favonia) [inline src/core/IDisjointSet.ml]: Now this operation is hidden because `subst` is made available.

**Comment by favonia** [comment]
I'm not very motivated to fix other files if we are going to embrace the new domain...

**Comment by jonsterling** [comment]
@favonia Then, I suggest we combine efforts / branches?

**Comment by favonia** [comment]
> @favonia Then, I suggest we combine efforts / branches?

Yes, just take the files.

**Comment by jonsterling** [comment]
These files are now integrated into #270

**Pull request closed without merging**